### DISCUSSION TITLE
Using both Build and Deploy Workflows

### DISCUSSION THREAD

**Discussion by Porter-dm** [community/community#37914] [Actions]
I am looking for advice on how to architect my builds and deployments in GitHub Actions.

For the build workflow, I want to build my project and save the output files. 

For the deployment workflow, I want to download the output file from a specific build, and then deploy it to a specific environment.

How would I accomplish this?  

Specifically:
- Where should I save the build output?  As a workflow artifact?  As a package?  As a workflow cache?
- In the deployment workflow, how do I specify the output file from a specific build/commit?

**Comment by airtower-luna** [comment]
> Where should I save the build output? As a workflow artifact? As a package? As a workflow cache?

A package is best if you want the build result to be available permanently. Artifacts are kept for a limited time, configurable with a maximum of 90 days (plenty if you just want to transfer data between jobs). Definitely *not* a cache, that's not meant for any use where you need targeted, reliable retrieval.

> In the deployment workflow, how do I specify the output file from a specific build/commit?

A common pattern is to have one *workflow* with separate *jobs* to build/test, and then deploy. You can put a condition on the deploy job (e.g. "run only on tags").

Of course this depends a lot on the conditions under which you want to deploy. If you want to control deployment completely separately from building you might need a very different approach, though in that case I'd ask if the two things even belong into the same repository. :slightly_smiling_face:

**Comment by Porter-dm** [comment]
I have over a dozen environments to deploy to.  For many of the environments, deployments should only happen when manually triggered.

This is why I was thinking it would make sense to have separate build and deployment workflows -- build workflow could run for every commit, while deployment workflows would be manually triggered.

Is there some better way of doing this?

**Comment by airtower-luna** [comment]
That sounds a bit like you have a software that will get deployed to various customer environments or something similar. In that case I'd probably have the software build "deploy" in the sense of uploading to a suitable package archive (or simply network storage).

Then you can have a separate automation that retrieves a selected package version and deploys it to a certain customer (or other) environment. I'd prefer a config file in Git over inputs, so it's easily reproducible (see GitOps for the high art of that), but that depends a bit on what you need.

**Comment by Porter-dm** [comment]
You are correct -- we are deploying to various customer and testing stacks, both internal and external.

> In that case I'd probably have the software build "deploy" in the sense of uploading to a suitable package archive (or simply network storage).
>Then you can have a separate automation that retrieves a selected package version and deploys it to a certain customer (or other) environment. 

That makes sense.  If I were to do this, is there a way to point the deployment worfklow to a specific build of the build workflow in order to get the specific package version uploaded by that build?

> I'd prefer a config file in Git over inputs, so it's easily reproducible (see GitOps for the high art of that), but that depends a bit on what you need.

Can you help me understand what you mean by this?

**Comment by airtower-luna** [comment]
> If I were to do this, is there a way to point the deployment worfklow to a specific build of the build workflow in order to get the specific package version uploaded by that build?

It depends on how you store the build results, of course, but assuming you use some kind of versioning yes. All package managers I'm familiar with offer a way to select a specific version, but the details vary. For development versions that need testing Actions artifacts (which have a limited lifetime) might be sufficient. Either way you'd then need to tell your deployment which version to select (think "deploy version 1.2.3 to customer A", or "deploy build artifacts from commit 1234abcd to test server B").

As for config files, imagine you have a config file that defines exactly how a certain deployment (or multiple ones) should be done. Again the details will vary depending on the exact tools you use (e.g. Ansible for deployments to bare metal servers or VMs, Helm for deployments to k8s). Example made up on the spot, YAML-style, but not for a specific tool, just to illustrate the idea:

```yaml
customer_name: Customer A
servers:
  - hostname: server1.example.com
    auth: ... # some secrets here (encrypted, or a reference to secure storage)
  - hostname: server2.example.com
    auth: ...
tool_version: 1.2.3
```

This would mean that a "Customer A" deployment will install version 1.2.3 of your tool to the servers on the list. Of course you could point your deployment tool of choice at the config file manually, but once you're that far why not automate? So if you push the config file, the deployment runs. If the customer then wants to update to 1.2.4 just make the change in the config, and push.

The general idea of GitOps is that you define the intended state (of a cluster, deployment, etc.) in config files that are versioned in Git, and a suitable tool (or tools) makes sure the config is applied. Obviously it's a bit more complex in practice, but I think it's a very elegant concept. :slightly_smiling_face:

**Comment by mickeygousset** [comment]
YOu could also use releases. Create a draft release with your deployment files in it. Then, when you have a release that you want to deploy, move it from draft to published and have a workflow that runs, grabs the files from the release, and deploys them.
